feat: Add persistent device database for offline/degraded mode operation

maxfield-allison opened the pull request:
## Summary

This PR implements the persistent device database foundation we discussed in #76 and #537. It enables govee2mqtt to gracefully degrade to LAN-only mode when Govee APIs are unavailable, rather than crashing.

## Problem

When all Govee APIs fail (authentication error, rate limiting, abnormal activity detection, or network issues) **and** the SQLite cache is empty or cleared, govee2mqtt crashes with the `ISSUE_76_EXPLANATION` error. This prevents LAN control even though LAN devices are reachable on the local network.

## Solution

A persistent JSON device database that:
- Stores device metadata (id, sku, name, room) learned from Platform and Undoc APIs
- Survives cache clears and container restarts (separate from SQLite cache)
- Enables graceful degradation to LAN-only mode
- Preserves device names for Home Assistant entity stability
- Human-editable JSON format as you requested

### Implementation Details

Following the architecture you outlined in #537:

**Storage (src/device_database.rs)**
```rust
#[derive(Serialize, Deserialize)]
pub struct DeviceDatabase {
    pub version: u32,
    pub devices: BTreeMap<String, PersistedDevice>,
}
```

- **Location**: `/data/devices.json` (HA addon) or `~/.cache/govee2mqtt/devices.json`
- **Atomic writes**: temp file + rename pattern for crash safety
- **User override fields**: `user_name` and `user_room` for future editing capability

**Startup Flow (src/commands/serve.rs)**
1. Detect startup mode: Fresh, Upgrade, or Normal
2. Load existing device database
3. Attempt API discovery
4. On API failure with cached data → continue with warnings
5. Populate in-memory state from database in degraded mode
6. Start LAN discovery

**Fallback Logic**
- API succeeds → database updated with latest metadata
- API fails + database exists → use cached data, log warnings
- API fails + no database → fail with ISSUE_76 (unchanged behavior for fresh installs)

## Changes

| File | Changes |
|------|---------|
| `src/device_database.rs` | **New** - Device database implementation |
| `src/commands/serve.rs` | Startup integration, fallback logic |
| `src/service/device.rs` | `cached_name`/`cached_room` fields |
| `src/service/state.rs` | Device database handle in state |
| `src/main.rs` | Module declaration |
| `Cargo.toml` | Dependencies (chrono, dirs, tempfile) |

## Testing

Tested on a Docker Swarm deployment with 7 Govee devices:

| Scenario | Result |
|----------|--------|
| Valid credentials | ✅ Database populated, all devices work |
| Invalid API key + SQLite cache exists | ✅ Cache returns stale data, service runs |
| **Invalid credentials + NO SQLite cache** | ✅ **Loads from devices.json, LAN control works** |
| Fresh install + no credentials | ✅ Fails as expected (no data to use) |

### Key Test Output (Scenario 3)
```
Platform API discovery failed: status 401 Unauthorized
Continuing with 7 cached devices from persistent database
Undoc API discovery failed: Incorrect user name or password
Continuing with 7 cached devices from persistent database
Populating device state from persistent database...
Loaded 7 devices into memory from persistent database
Running in degraded mode - device metadata may be stale
LAN-capable devices should still be controllable
Starting LAN discovery
...
Using LAN API to set Frame Front Accent power state
```

## Non-Breaking

- ✅ All existing tests pass (`cargo test`)
- ✅ Existing SQLite cache behavior unchanged
- ✅ Database is purely an additive resilience layer
- ✅ Fresh installs without API access still fail appropriately

## Future Work (Not in this PR)

This is intentionally a focused foundational PR. Future enhancements could include:

1. **Entity ID stability** - Lock entity names on first registration
2. **Web UI management** - View/edit device database from the web interface  
3. **Per-device API preferences** - Override which API to use per device
4. **Explicit LAN IP configuration** - Manual IP entry for VLAN/discovery issues

## Related

- Resolves foundational work for #76
- Supersedes #537 (which was rejected due to scope issues)

---

Happy to address any feedback or make adjustments to the approach!

maxfield-allison committed:
feat: Add persistent device database for offline/degraded mode operation

This implements the device database foundation that Wez requested for
resolving Issue #76 (LAN-only fallback when APIs are unavailable).

## Problem Solved

When all Govee APIs fail (authentication error, rate limiting, network
issues) and the SQLite cache is empty/cleared, govee2mqtt would crash
with the ISSUE_76_EXPLANATION error. This prevented LAN control even
though LAN devices were reachable.

## Solution

A persistent JSON device database that:
- Stores device metadata (id, sku, name, room) learned from APIs
- Survives cache clears and container restarts
- Enables graceful degradation to LAN-only mode
- Preserves device names for Home Assistant entity stability

## Changes

### New: src/device_database.rs
- JSON storage at /data/devices.json (HA addon) or ~/.cache/govee2mqtt/
- Atomic writes (temp file + rename) for crash safety
- StartupMode detection: Fresh, Upgrade, or Normal
- User override fields for future editing capability

### Modified: src/commands/serve.rs
- Load device database on startup
- Fallback to cached devices when APIs fail (with warnings)
- Populate in-memory state from database in degraded mode
- Update database on successful API/LAN discovery

### Modified: src/service/device.rs
- Added cached_name/cached_room fields
- name() and room_name() fall back to cached values

## Tested Scenarios

1. Valid credentials: Database populated, all devices work
2. Invalid creds + SQLite cache: Cache returns stale data
3. Invalid creds + NO cache: Loads from devices.json, LAN works
4. Fresh install + no creds: Fails as expected (no data)

## Non-Breaking

- All existing tests pass
- Existing SQLite cache behavior unchanged
- Database is purely additive resilience layer

wez commented on src/device_database.rs:
```diff
@@ -0,0 +1,627 @@
+    /// Load database from disk, or create empty if not exists
+    pub fn load(path: &Path) -> anyhow::Result<Self> {
+        if path.exists() {
+            let contents = fs::read_to_string(path)?;
```
FWIW, this `if file-exists { read file }` pattern is an example of a [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) issue.  It is possible (although admittedly unlikely in this use case) for something to delete the file between the time we look for it and the time we try to open it.

The "right" way to deal with this is to attempt to read the file, and convert the file-not-found error into the default fallback case.

Something like:

```rust
match std::fs::read_to_string(path) {
   Ok(contents) => { ... }
   Err(err) if err.kind() == std::io::ErrorKind::NotFound => { ... use default ... }
   Err(err) => return Err(err),
}
```

This has the added micro-optimizational benefit of saving an extraneous `stat(2)` system call because we don't need to check for existence explicitly first; just let the kernel tell us as part of reading the file.

wez commented on src/device_database.rs:
```diff
@@ -0,0 +1,627 @@
+        }
+
+        // Write to temporary file in same directory
+        let temp_path = path.with_extension("json.tmp");
```
From a defensive programming perspective, this is prone to pathological problems if, for example, `json.tmp` already exists and is some broken state (permissions are bad, or some other weird filesystem error), and also to things like [symlink races](https://en.wikipedia.org/wiki/Symlink_race).  Handling all the edge cases 100% is tricky but also not strictly required here.

For the sake of simplicity and getting more robustness "for free", I'd recommend that we adopt [NamedTempFile::new_in](https://docs.rs/tempfile/latest/tempfile/struct.NamedTempFile.html#method.new_in) here to sidestep these.  Something like:

```rust
let file = NamedTempFile::new_in(path.parent()?)?;
file.write_all(contents.as_bytes())?;
file.sync_all()?;
file.persist(path)?;
```

https://docs.rs/tempfile/latest/tempfile/struct.NamedTempFile.html#method.persist

Note that the `persist` call is also potentially prone to a symlink race if someone replaces the parent directory with a symlink; I'm not going to die on that particular hill; this suggestion is already good enough!

wez commented on src/device_database.rs:
```diff
@@ -0,0 +1,627 @@
+        }
+
+        // Atomic rename
+        fs::rename(&temp_path, path)?;
```
re: comments above, if the rename fails for some reason here, `json.tmp` will remain in the filesystem.  Using the tempfile module, the destructor will clean it up for us, so that's another point in its favor.

wez commented on src/device_database.rs:
```diff
@@ -0,0 +1,627 @@
+            .unwrap_or(false)
+        {
+            return addon_path;
+        }
```
This could be a bit funky because it will use `/data/devices.json` if run on any system that happens to have `/data` on the filesystem.

I think it would be best overall (but sadly more fussy in the code) to add an optional command line argument for this to `ServeCommand` (clap's `arg` stuff should make this easy) and then use `opt_env_var` to facilitate the env var configuration mode.

Then in `addon/run.sh` we can set the command line argument or env var explicitly for the addon.

wez commented on src/commands/serve.rs:
```diff
@@ -176,64 +225,158 @@ impl ServeCommand {
+            );
+        }
+
+        state.set_device_database(device_db.clone()).await;
```
My thinking is that right here is a good place to unconditionally enumerate the devices from the persistent database and apply them to the device state; essentially the logic that you have below in the degraded mode case around line 363.

I think that we can safely unconditionally do that here and that it will simplify some logic later on: that way we always know about the devices that existed from the last run, and the discovery logic can proceed to augment it.  TBH, that augmentation is likely to be a NOP--it should be the same essential information in the common case anway.

wez commented on src/commands/serve.rs:
```diff
@@ -176,64 +225,158 @@ impl ServeCommand {
+                        log::warn!(
+                            "Continuing with {} cached devices from persistent database",
+                            cached_device_count
+                        );
```
This particular notice and the logic that produces it can be simplified; if we're unconditionally loading the device database where I indicate above, then what we can do is have a block after we're done with all of the discovery below where we check to see if we have any API updates.

As a rough sketch of what I mean, the logic in here would be:

1. Get the current time
2. Load the persistent database, apply to device state
3. Perform the various discovery methods, which update the `api_info` field in the persisted device state
4. Iterate the persisted device state, if a device doesn't have a `last_sync` newer than the time we got in step 1 then we didn't see it during discovery and can print out a note that we're using cached data for it.

This saves us from juggling multiple flags and multiple different discovery states here in this block, which should reduce the number of branches and complexity of the code.

wez commented on src/commands/serve.rs:
```diff
@@ -176,64 +225,158 @@ impl ServeCommand {
+                            "Error during initial undoc API discovery: {err:#}\n\
+                            No cached device data available to fall back to.\n\
+                            {ISSUE_76_EXPLANATION}"
+                        );
```
similar comments as the case above: I think we can move the notice to happen more centrally after all discovery is done, and remove the ISSUE_76 error case.

wez reviewed (commented):
Thank you again for this! Sorry for the delay in responding; it's a busy time of year!

maxfield-allison committed:
Address review feedback from wez

- Avoid unnecessary clone in device iteration (use reference)
- Use defensive slicing with .get().unwrap_or() to prevent panics
- Fix TOCTOU race condition in database load
- Use NamedTempFile for robust atomic writes
- Rename cached_name/cached_room to name/room (database is source of truth)
- Restructure discovery flow: always load database first
- Remove ISSUE_76 crash - fresh installs use generated names

maxfield-allison commented:
one more commit incoming shortly to address the remaining items from review. or i can close and push one clean if you prefer, apologies.

maxfield-allison commented:
need to correct database saves on LAN status query

maxfield-allison committed:
Fix: Remove excessive database saves on LAN status updates

The save was being called on every LAN status response, causing multiple
writes per minute. Database saves should only happen after API enumeration,
not on every status update.

maxfield-allison committed:
docs: Remove stale env var reference from module documentation

maxfield-allison committed:
Merge branch 'wez:main' into feature/lan-fallback-device-database

inventor7777 commented:
Nice thinking.

that's another pain point I have with the current implementation - it doesn't pull everything (I have hundreds of DIYs and it only pulls half or less)

maxfield-allison committed:
Merge upstream/main: resolve Cargo.toml conflict with newer deps

mercertom commented:
When can we get this? I can't use adaptive lighting until this is fixed. My service just insta-pings over the daily or minute or second or hour limit, and then my lights are stuck at some arbitrary brightness.
